packages/engine-server/src/drivers/file/storev2.ts

```diff
@@ -649,161 +649,149 @@ export class FileStorage implements DStore {
+            // And if this used to be a frontmatter tag, the alias being undefined will force it to be removed because a frontmatter tag can't point to something outside of tags hierarchy.
+            alias = undefined;
+          }
+          // for user tag links, we'll have to regenerate the alias
+          if (newLoc.fname.startsWith(USERS_HIERARCHY)) {
+            const fnameWithoutTag = newLoc.fname.slice(USERS_HIERARCHY.length);
+            alias = `@${fnameWithoutTag}`;
+          } else if (oldLink.from.fname.startsWith(USERS_HIERARCHY)) {
+            // If this used to be a user tag but no longer is, the alias is like `@foo.bar` and no longer makes sense.
+            alias = undefined;
+          }
+          // Correctly handle header renames in references with range based references
+          if (
+            oldLoc.anchorHeader &&
+            link.type === "ref" &&
+            isNotUndefined(oldLink.from.anchorHeader) &&
+            oldLink.from.anchorHeader.indexOf(":") > -1 &&
+            isNotUndefined(newLoc.anchorHeader) &&
+            newLoc.anchorHeader.indexOf(":") === -1
+          ) {
+            // This is a reference, old anchor had a ":" in it, a new anchor header is provided and does not have ":" in it.
+            // For example, `![[foo#start:#end]]` to `![[foo#something]]`. In this case, `something` is actually supposed to replace only one part of the range.
+            // Find the part that matches the old header, and replace just that with the new one.
+            let [start, end] = this.referenceRangeParts(
+              oldLink.from.anchorHeader
+            );
+            if (start === oldLoc.anchorHeader) start = newLoc.anchorHeader;
+            if (end === oldLoc.anchorHeader) end = newLoc.anchorHeader;
+            newLoc.anchorHeader = `${start}:#${end}`;
+          }
+          const newBody = LinkUtils.updateLink({
+            note,
+            oldLink,
+            newLink: {
+              ...oldLink,
+              from: {
+                ...newLoc,
+                anchorHeader: newLoc.anchorHeader || oldLink.from.anchorHeader,
+                alias,
               },
-            });
-            _n.body = newBody;
-            return _n;
-          },
-          _n
-        );
-        // const resp = await MDUtilsV4.procTransform(
-        //   { engine: this.engine, fname: n.fname, vault: n.vault },
-        //   { from: oldLoc, to: newLoc }
-        // ).process(_n.body);
-        n.body = noteMod.body;
-        n.tags = noteMod.tags;
-        return n;
-      })
-    ).catch((err) => {
-      this.logger.error({ err });
-      throw new DendronError({ message: " error rename note", payload: err });
+            },
+          });
+          _n.body = newBody;
+          return _n;
+        },
+        _n
+      );
+      n.body = noteMod.body;
+      n.tags = noteMod.tags;
+      const shouldChange = !(
+        n.body === noteMod.body && n.tags === noteMod.tags
```

Comment:
small nitpick: could compare content hash instead

Reply:
we could, but `noteMod` that we are comparing `n` with doesn't have a contentHash to start with (file2Note doesn't calculate hash). We can pass in `calculateHash` to `file2Note` and get it when we are reading `_n`, but then we also have to re-calculate the content hash of `noteMode` right before we check if there's a change. I think it would be better to just use what we already know might have changed to check here instead of adding hash calculation to an operation that can possibly span out to many notes.